Httponly cookie

SpencerWightman opened the pull request:
- add a `SESSION_SECRET` to your `.env`
- local storage references should be gone
- new `express-session` and `connect-redis` libraries
- `express-session` allows us to set HTTPOnly cookies in socket.io middleware
- `express-session` can sync client session data with cache or db: shared between socket servers
- right now it syncs with only Redis but the `CustomStore` logic in `index.ts` should be ready for Dynamo logic to be plugged in -- need time difference conditional to determine when to read/set session data from Redis vs. from Dynamo
- `twineTS` is set to `currentTimestamp` on first connection, and updated when a new or rewind message is received by the client: `updateSessionTS` event emits from within the `message` event listener
- `twineID/RC/TS` persists across server restart, unintentional disconnect, intentional disconnect
- `twineRC` is a boolean in the session used on `line 101` in `socketServices` to determine a reconnect 
- to prevent messages sent to a room before the client joined the room from being sent to them on reconnect, the value for the `roomName` field in the Redis `subscribedRooms` hash is now the timestamp of when the client joined the room (before it was just the name of the key)
  - that 'room join' timestamp is compared to the `twineTS` timestamp (generated on each message received) in `redisService` `line 58`: the greater value is used to determine which messages were missed

**notes**
- still have to confirm cookie `expiry` works as expected
  - we could set cookies to expire at connection state recovery max (24hrs)
- there is another solution that uses local storage for `twineTS`
- we should consider checking to see if a message was missed before emitting missed messages
- this kind of type assertion `socket.twineID = (socket.request as any).session?.twineID as string;` should be cleaned up eventually
- prevented Redis from consistently returning the correct missed messages + 1 message already seen by adding +1 to the timestamp passed to `processSubscribedRooms`
  - the timestamp associated with each published message is generated in `expressServices` `publish`: it is used in the Redis message sorted set and also to update the client's `twineTS` -- idea is that because the timestamp is set when a message is first received by our HTTP API, it will be stored as the same value for every client that receives the message (latency individual to each client does not matter), and also match in Redis -- but maybe because it's a perfect match between client and Redis, there was the off by 1 error (`zrangebyscore` inclusive of the minimum?)
- maybe a good time to create edge case tests
  - client first time connect > no messages received > disconnects > misses messages > reconnects > receives missed messages
  - messages published to room client is not in > client joins rooms > client reconnects > client should not(?) receive missed messages
- `express-session` gives an ID to each session that in a way makes `twineID` redundant, but `twineID` works well to check if it's a reconnect or not -- doing that with the `express-session` ID might be complicated

LeeyaD commented:
Hey Spencer, 

Great work with the HTTPonly Cookie! There's just 2 things I wanted to mention:

It looks like I'm unable to connect to the server without an existing `SESSION_SECRET` in the `.env` file, is this how it's supposed to work for now?

If it's not, here's more detail on what I did, to hopefully help with debugging:
- I started the server and received the following warning: `express-session deprecated req.secret; provide secret option file:/...ws_server/dist/index.js:56:27`, which appears to be referencing `#session` on lines 67-79 of `index.ts`.
- Since it was just a warning, I decided to move on to just test out the app, going to `localhost:3002` but my browser was stuck on polling and only ever received a`400` each time. The terminal didn't show any errors.
- I then went to `localhost:3001` expecting to receive the usual 'Nice work' string, but this error was raised instead: `Error: secret option required for sessions`

Hardcoding a `SESSION_SECRET` in `.env` solves this issue, but I'm can't remember if this is what we wanted for now or if we were expecting it to generate/update on it's own.

Also, it looks like the server restart issue persists. `twineTS` only updates when the `disconnecting` event fires and a server restart doesn't fire that event, so when a user reconnects they receive all messages (incl. ones they previously received while connected).

SpencerWightman commented:
Good catch. I added a line about `SESSION_SECRET`. I think `.env` works for local development because that's what we've been using for secrets. Will probably change when everything is on AWS but I haven't looked into that yet. Updated the PR details: `twineTS` persists through server restart now (in my tests) but needs to be tweaked for at least one edge case.

ChristopherBrum commented:
Fantastic work, Spencer!
